helm-core.el

```diff
@@ -5121,16 +5112,17 @@ without recomputing them, it should be a list of lists."
              (member (assoc-default 'name source) helm-source-filter))
          (>= len
              (helm-aif (assq 'requires-pattern source) (or (cdr it) 1) 0))
-         ;; Entering repeatedly these strings (*, ?) takes 100% CPU
-         ;; and hang emacs on MacOs preventing deleting backward those
-         ;; characters (Bug#1802). Update: it seems it is no more true,
-         ;; thus this affect bug#2423, so let's remove this for now.
-         ;; (not (string-match-p "\\`[*]+\\'" helm-pattern))
-         ;; These incomplete regexps hang helm forever
-         ;; so defer update. Maybe replace spaces quoted when using
-         ;; multi-match.
-         (not (member (replace-regexp-in-string "\\s\\ " " " helm-pattern)
-                      helm-update-blacklist-regexps)))))
+         ;; bugs were mentioned here before: Bug#1802, bug#2423
+         ;; must be fixed already by 'helm-re-search-forward'
+         ;;
+         ;; short patterns that are alone not expected to return any results,
+         ;; ignore in order to not erase Helm buffer on such patterns.
+         (not
+          (let ((split-string-default-separators "[ \f\t\n\r\v]+")
+                (subpatterns (string-split helm-pattern)))
+            (seq-find
+             (lambda (p) (member p '("!" "!^" "?" "@!")))
+             subpatterns nil))))))
```

Comment:
I avoid using seq in Helm for many reasons:
- compatibility with other Emacs versions (different seq versions depending Emacs versions)
- Performances
- Readability
- seq is not needed unless the argument passed to it may be of different kind (list, vector or whatever)

I prefer using cl-loop:

```elisp
(cl-loop for p in (helm-mm-split-pattern helm-pattern)
                  never (member p '("!" "!^" "?" "@!")))
``` 
Using never avoid the not on top.

Reply:
> I prefer using cl-loop:

That's better. Checking.